src/Symfony/Component/VarExporter/Instantiator.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\VarExporter;
+
+use Symfony\Component\VarExporter\Internal\Hydrator;
+use Symfony\Component\VarExporter\Internal\Registry;
+
+/**
+ * A utility class to create objects without calling their constructor.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class Instantiator
+{
+    /**
+     * Creates an object and sets its properties without calling its constructor nor any other methods.
+     *
+     * For example:
+     *
+     *     // creates an empty instance of Foo
+     *     Instantiator::instantiate([Foo::class => []]);
+     *
+     *     // creates a Foo instance and sets one of its public, protected or private properties
+     *     Instantiator::instantiate([Foo::class => ['propertyName' => $propertyValue]]);
+     *
+     *     // creates a Foo instance and sets a private property defined on its parent Bar class
+     *     Instantiator::instantiate([
+     *         Foo::class => [],
+     *         Bar::class => ['privatePropertyName' => $propertyValue],
+     *     ]);
+     *
+     * Instances of ArrayObject, ArrayIterator and SplObjectHash can be created
+     * by using the special "\0" property name to define their internal value:
+     *
+     *     // creates an SplObjectHash where $info1 is attached to $obj1, etc.
+     *     Instantiator::instantiate([SplObjectStorage::class => ["\0" => [$obj1, $info1, $obj2, $info2...]]]);
+     *
+     *     // creates an ArrayObject populated with $inputArray
+     *     Instantiator::instantiate([ArrayObject::class => ["\0" => $inputArray]]);
+     *
+     * @param array $propertiesByClass The properties to set on the object, keyed by the class to bind to
+     *                                 when setting them. The first key of the array defines the class
+     *                                 of the returned instance.
+     *
+     * @return object An instance of the class defined by the first key of $propertiesByClass
+     *
+     * @throws \ReflectionException When the instance cannot be created
+     */
+    public static function instantiate(array $propertiesByClass)
+    {
+        if (!$propertiesByClass) {
+            throw new \ReflectionException('No class provided.');
+        }
+
+        foreach ($propertiesByClass as $class => $properties) {
+            break;
+        }
+
+        try {
+            $reflector = Registry::$reflectors[$class] ?? Registry::getClassReflector($class);
+        } catch (\Exception $e) {
+            if (0 === strpos($e->getMessage(), 'Serialization of ')) {
+                $e = new \ReflectionException(substr_replace($e->getMessage(), 'Instanti', 0, 8));
+            }
+
+            throw $e;
+        }
+
+        if (Registry::$cloneable[$class]) {
```

Comment:
Why not use `doctrine/instantiator` directly here? Would save a lot of headaches and duplicate efforts, and already performs these checks at the cost of one method call. Are there test cases that we need to port over there too?

Reply:
The headaches are already past stories :)
The `VarExporter::export()` need access to internal details when generating its output, so this is already solved concern.